### PR TITLE
ENH: Add require argument to load() to accept version specifiers

### LINKED ISSUE #13: Optional package features
I want to run a couple ideas past this group, which is mostly targeted at deciding whether it makes more sense to help implement features here versus update our similar feature to use lazy loading under the hood. The use case in mind is [optional (extra) dependencies](https://packaging.python.org/en/latest/specifications/core-metadata/#provides-extra-multiple-use), and see the next (collapsed) section for details.

The features I'm proposing are are:

1) Version constraints (I'm thinking [`packaging.requirements`](https://packaging.pypa.io/en/latest/requirements.html) arbitrary strings, but minimum versions are probably >90% of the use case)
2) Boolean "found package" return value

Here is a quick API proposal to demonstrate:

```Python
import lazy_loader as lazy
h5py, have_h5py = lazy.load_requirement('h5py>=3.0')
```

Here `h5py` would be a lazy module, which will fail if either `h5py` is not installed or its version does not satisfy `>=3.0`. `have_h5py` will be a boolean that can be used to determine whether accessing attributes of `h5py` will fail without attempting the load.

I think this would in practice require adding `importlib_metadata` (for Python <3.9) and `packaging` to the dependencies, so it might be out of scope if you're trying to keep this absolutely minimal.

---

<details>
<summary>Click for some details about what I would use this API to replace</summary>

We've been doing something similar to `lazy.load` in [`nibabel.optpkg.optional_package`](https://nipy.org/nibabel/reference/nibabel.optpkg.html#optional-package) for a while, that looks like this:

```Python
from nibabel.optpkg import optional_package
h5py, have_h5py, setup_module = optional_package('h5py', min_version='3.0')
```

`h5py` is either the fully loaded `h5py` module or a "tripwire" object that will fail when you attempt to access an attribute. So it is (currently) only lazy in the sense that an `ImportError` does not occur immediately (and will be replaced by a more appropriate informative error message). But from a namespace standpoint, `lazy.load('h5py')` and `optional_package('h5py')[0]` are interchangeable.

Importantly for the optional dependency use case, an installed version that is lower than the required version is treated as a failure to import. This can come up periodically, and a dependency mismatch is only relevant to the user or downstream tool if they are accessing h5py functionality in nibabel which might require a higher version. (The error message, of course, explains the issue.)

The other features are `have_<module>`, which is a simple `bool`. I personally mainly use it for tests like `@unittest.skipUnless(have_h5py, reason="Needs h5py")`, but you can imagine other cases when it's more convenient than accessing the module and allowing/catching an exception.

Finally, `setup_module` is just a convenience that will raise a `SkipTest()` exception so a module that can't be tested at all without the dependency will be skipped by tools like `nose` or `pytest`. (Often we will use `h5py, have_h5py, _ = optional_package('h5py'), if we don't want to skip the entire module.) This is almost entirely unused even internally, so I'm not proposing to add it here.

</details>

### PR DESCRIPTION
**Pull request by effigies**
This PR adds an optional `require` keyword argument to `load()` that accepts `requirements.txt`-style version specifiers.

```python
np = lazy.load("numpy", require="numpy >=1.24")
```

The effect is to make the import fail if the dependency exists but the version is not satisfied.

Closes #13.

---- 

<details><summary>Original text</summary>

Found a little time tonight to take a first stab at implementing `lazy.load_requirement()`, a variant of `lazy.load()` that accepts a `requirements.txt`-style requirement specifier. This would allow one to write:

```Python
nibabel, have_nibabel = lazy.load_requirement('nibabel >= 4.2')
yaml, have_yaml = lazy.load_requirement('pyyaml >= 6.0', 'yaml')
stats, have_scipy = lazy.load_requirements('scipy >= 1.9', 'scipy.stats')
```

The two-argument form is for when module names do not match distribution names or you want to lazily import a submodule while constraining the base module version.

I'll get to tests when I can, but might as well see if this much complexity is within-scope. I doubt it can be done much simpler.

</details>

### CODE REVIEW

**Comment by stefanv** [comment]
I think this could work!

Would we be able to re-use some existing code by making the interface `lazy.load(spec, requires=...)`, instead of defining an entirely new function?

Adding the new function is also fine, although I would swap the order of the arguments `scipy.stats, scipy >= 1.9`.

File: lazy_loader/__init__.py
Comment (stefanv) [inline lazy_loader/__init__.py:184]: What does this comment mean?

File: lazy_loader/__init__.py
Comment (stefanv) [inline lazy_loader/__init__.py:196]: What does this refer to?

File: lazy_loader/__init__.py
Comment (stefanv) [inline lazy_loader/__init__.py:231]: Is `have_module` used anywhere?

**Review by stefanv — COMMENTED** [review]
This looks good already!

File: lazy_loader/__init__.py
Comment (effigies) [reply lazy_loader/__init__.py:184]: Ah. Initially I'd used `load()` to lazily load `packaging.requirements` at the module root, but it seems better not to add it to `sys.modules` unless calling code actually makes use of `packaging.requirements`. So "old style lazy loading" just means import at function runtime.

File: lazy_loader/__init__.py
Comment (effigies) [reply lazy_loader/__init__.py:196]: I suspect a common error will be to do something like `yaml = load('yaml', 'yaml >= 5')`. The `yaml` module is provided by the `pyyaml` distribution, so it needs to be `yaml = load('yaml', 'pyyaml >= 5')`.

This is necessary because not all modules provide a `__version__`, but all distributions must provide version metadata, and that's what's retrieved by `importlib.metadata.version`.

File: lazy_loader/__init__.py
Comment (effigies) [reply lazy_loader/__init__.py:231]: Not internally. I thought it would be a nice addition to the API, as I will often do something like:

```Python
# Was: h5py, have_h5py, _ = optional_package('h5py')
h5py = lazy.load('h5py')
have_h5py = lazy.have_module(h5py)

needs_h5py = unittest.skipUnless(have_h5py, 'Test requires h5py')

@needs_h5py
def test_mything():
    ...
```

I included it here because I initially had `load_requirement() -> tuple[module, boolean]`, so this replaced that second return value. Could be a separate PR to keep this clean, or I can just do it downstream, myself. The lazy loading and deferred errors are the important things to me.

**Comment by effigies** [comment]
I was curious about how heavy these imports are, so I moved importing `inspect`, `packaging.requirements` and `importlib.metadata` inside the functions that need them.

<details><summary>Click for diff</summary>

```diff
diff --git a/lazy_loader/__init__.py b/lazy_loader/__init__.py
index 016c26d..f3427c1 100644
--- a/lazy_loader/__init__.py
+++ b/lazy_loader/__init__.py
@@ -7,16 +7,10 @@ Makes it easy to load subpackages and functions on demand.
 import ast
 import importlib
 import importlib.util
-import inspect
 import os
 import sys
 import types
 
-try:
-    import importlib_metadata
-except ImportError:
-    import importlib.metadata as importlib_metadata
-
 __all__ = ["attach", "load", "attach_stub"]
 
 
@@ -181,9 +175,14 @@ def load(fullname, *, require=None, error_on_import=False):
     if not have_module:
         not_found_message = f"No module named '{fullname}'"
     elif require is not None:
         # Old style lazy loading to avoid polluting sys.modules
         import packaging.requirements
 
+        try:
+            import importlib_metadata
+        except ImportError:
+            import importlib.metadata as importlib_metadata
+
         req = packaging.requirements.Requirement(require)
         try:
             have_module = req.specifier.contains(
@@ -202,6 +201,8 @@ def load(fullname, *, require=None, error_on_import=False):
     if not have_module:
         if error_on_import:
             raise ModuleNotFoundError(not_found_message)
+        import inspect
+
         try:
             parent = inspect.stack()[1]
             frame_data = {
```

</details>

Then I profiled loading `lazy_loader` and each of these requirements in turn.

```console
$ python -X importtime -c "import lazy_loader; import inspect; import packaging.requirements; import importlib.metadata" 2>&1 | awk '/[0-9e] \| [a-z]/' 
import time: self [us] | cumulative | imported package
[...]
import time:       205 |       4984 | lazy_loader
import time:      1067 |       3710 | inspect
import time:       156 |      19275 | packaging.requirements
import time:       860 |      12247 | importlib.metadata
```

`packaging.requirements` and `importlib.metadata` are each 2-4x slower to load than `lazy_loader` itself, and only needed if this feature is called for. That said, we're talking about a one-time import cost of ~30-35ms. I'll leave it to you to decide if we should import everything at the module level or defer some until needed.

**Comment by stefanv** [comment]
I think this is the kind of library where that type of optimization makes perfect sense.

**Comment by effigies** [comment]
Rebased, smoothed down ruff's feathers about how complex `load()` was getting, and removed the `have_module()` function as out-of-scope.

This is ready for review, whenever it's convenient.

**Review by dschult — COMMENTED** [review]
This looks like a reasonable feature and it seems set up well. I don't know the `mock` unittest construct details but the intent seems good. I put a few comments below as suggestions.

File: lazy_loader/__init__.py
Comment (effigies) [reply lazy_loader/__init__.py:103]: > Do you want a default message?

I found it was easier to be more explicit about setting the message in the branching logic. I can rework with a default, if preferred.

> Also, why is message after *args in the function sig instead of before? For backward compat?

I figured keyword-only would be clearer, but I'm okay with any signature. Let me know if you'd like me to change this.

Comment (dschult) [reply lazy_loader/__init__.py:103]: Yes, I see what you mean and agree that dealing with the message is better handled in the branching logic.

**Review by dschult — APPROVED** [review]
I approve this PR (subject to the latest changes passing tests once the testing workflows  are approved). 
Thanks!

**Review by stefanv — APPROVED** [review]

**Comment by stefanv** [comment]
Thank you very much @effigies, for your contribution and your patience while we decided how to handle this.

**Pull request merged**
